src/common/valtext.cpp

```diff
@@ -186,143 +157,228 @@ bool wxTextValidator::Validate(wxWindow *parent)
-bool wxTextValidator::TransferToWindow()
+bool wxTextValidatorBase::TransferToWindow()
 {
-    if ( m_stringValue )
+    if ( m_string )
     {
         wxTextEntry * const text = GetTextEntry();
         if ( !text )
             return false;
 
-        text->SetValue(*m_stringValue);
+        text->SetValue(*m_string);
     }
 
     return true;
 }
 
 // Called to transfer data to the window
-bool wxTextValidator::TransferFromWindow()
+bool wxTextValidatorBase::TransferFromWindow()
 {
-    if ( m_stringValue )
+    if ( m_string )
     {
         wxTextEntry * const text = GetTextEntry();
         if ( !text )
             return false;
 
-        *m_stringValue = text->GetValue();
+        *m_string = text->GetValue();
     }
 
     return true;
 }
 
-// IRIX mipsPro refuses to compile wxStringCheck<func>() if func is inline so
-// let's work around this by using this non-template function instead of
-// wxStringCheck(). And while this might be fractionally less efficient because
-// the function call won't be inlined like this, we don't care enough about
-// this to add extra #ifs for non-IRIX case.
-namespace
+wxString wxTextValidatorBase::IsValid(const wxString& str) const
 {
-
-bool CheckString(bool (*func)(const wxUniChar&), const wxString& str)
-{
-    for ( wxString::const_iterator i = str.begin(); i != str.end(); ++i )
+    for ( wxString::const_iterator i = str.begin(), end = str.end();
+          i != end; ++i )
     {
-        if ( !func(*i) )
-            return false;
+        if ( !IsValid(*i) )
+            // N.B. this format string should contain exactly one '%s'
+            return _("'%s' contains invalid character(s)!");
     }
 
-    return true;
+    return wxString();
 }
 
-} // anonymous namespace
 
-wxString wxTextValidator::IsValid(const wxString& val) const
+void wxTextValidatorBase::SetCharIncludes(const wxString& chars)
 {
-    // wxFILTER_EMPTY is checked for in wxTextValidator::Validate
-
-    if ( HasFlag(wxFILTER_ASCII) && !val.IsAscii() )
-        return _("'%s' should only contain ASCII characters.");
-    if ( HasFlag(wxFILTER_ALPHA) && !CheckString(wxIsalpha, val) )
-        return _("'%s' should only contain alphabetic characters.");
-    if ( HasFlag(wxFILTER_ALPHANUMERIC) && !CheckString(wxIsalnum, val) )
-        return _("'%s' should only contain alphabetic or numeric characters.");
-    if ( HasFlag(wxFILTER_DIGITS) && !CheckString(wxIsdigit, val) )
-        return _("'%s' should only contain digits.");
-    if ( HasFlag(wxFILTER_NUMERIC) && !wxIsNumeric(val) )
-        return _("'%s' should be numeric.");
-    if ( HasFlag(wxFILTER_INCLUDE_CHAR_LIST) && !ContainsOnlyIncludedCharacters(val) )
-        return _("'%s' doesn't consist only of valid characters");
-    if ( HasFlag(wxFILTER_EXCLUDE_CHAR_LIST) && ContainsExcludedCharacters(val) )
-        return _("'%s' contains illegal characters");
-
-    return wxEmptyString;
+    wxASSERT( HasFlag(wxFILTER_INCLUDE_CHAR_LIST) );
```

Comment:
I think this should be a `wxCHECK_RET`. And then the redundant `HasFlag()` check below should be removed.